### ISSUE TITLE
[Bug]: aws_s3_bucket_lifecycle_configuration supports non-functional prefix filters

### ISSUE THREAD

**Issue by akloss-cibo** [hashicorp/terraform-provider-aws#28770]
### Terraform Core Version

1.3.7

### AWS Provider Version

4.49.0

### Affected Resource(s)

* aws_s3_bucket_lifecycle_configuration

### Expected Behavior

Terraform should refuse to apply this configuration.

### Actual Behavior

Terraform created a lifecycle configuration applying to the entire bucket even though the `prefix` argument is supplied.  While `prefix` is deprecated, it is a least astonishment (to put it mildly) to have a lifecycle configuration apply to the entire bucket when it was clearly intended to only apply to a prefix.

This also shows up as "permanent" drift, see output below.

### Relevant Error/Panic Output Snippet

```shell
# aws_s3_bucket_lifecycle_configuration.com-cibo-weather will be updated in-place
  ~ resource "aws_s3_bucket_lifecycle_configuration" "bucket" {
        id     = "some-bucket"
        # (1 unchanged attribute hidden)

      ~ rule {
            id     = "delete stuff"
          + prefix = "myprefix"
            # (1 unchanged attribute hidden)

            # (2 unchanged blocks hidden)
        }

        # (3 unchanged blocks hidden)
    }
```

### Terraform Configuration Files

```terraform
resource "aws_s3_bucket" "some-bucket" {
  bucket = "some-bucket"
}

resource "aws_s3_bucket_lifecycle_configuration" "some-bucket" {
  bucket = aws_s3_bucket.some-bucket.id

  rule {
    id     = "delete prefix stuff"
    status = "Enabled"
    prefix = "prefix"
    expiration {
      days = 1
    }
  }
}
```

### Steps to Reproduce

Apply the above configuration (with `"some-bucket"` changed to something unique).  `plan` it again and see that the filter isn't applied to the bucket.

### Would you like to implement a fix?

None

**Comment by justinretzolk** [comment]
Hey @akloss-cibo 👋 Thank you for taking the time to raise this! I've tried a few times with different provider and Terraform Core versions, but I've so far not been able to reproduce this. Is there anything different about your configuration when compared to the provided example? I'm wondering about things like interpolations vs. literal strings, etc.

If it's possible to supply them, debug logs (redacted as needed) may be helpful as well.

**Comment by akloss-cibo** [comment]
I can confirm your observation... my pruned example doesn't exhibit the behavior.  My real configuration still does and I'm experimenting with creating a good example.  In doing so, I just got this somewhat cryptic message:
```
error updating S3 Bucket Lifecycle Configuration (some-bucket-demo-terraform-28770): InvalidRequest: Base level prefix cannot be used in Lifecycle V2, prefixes are only supported in the Filter.
```
Is there some subtle API versioning going on inside the provider?  I'll keep experimenting, but I wanted to let you know ASAP.

**Comment by akloss-cibo** [comment]
OK, I figured out how to actually reproduce.  Apply this configuration (with the middle rule commented out):
```
resource "aws_s3_bucket" "some-bucket" {
  bucket = "some-bucket-demo-terraform-28770"
}

resource "aws_s3_bucket_lifecycle_configuration" "some-bucket" {
  bucket = aws_s3_bucket.some-bucket.id

  rule {
    abort_incomplete_multipart_upload {
      days_after_initiation = 7
    }
    id     = "Abort incomplete Multipart Uploads after 7 days"
    status = "Enabled"
  }

  */
  rule {
    id     = "delete stuff"
    prefix = "stuff"
    expiration {
      days = 1
    }
    status = "Enabled"
  }
  */

  rule {
    id = "expire objects"
    noncurrent_version_expiration {
      noncurrent_days = 45
    }
    expiration {
      expired_object_delete_marker = true
    }
    status = "Enabled"
  }
}
```

And then remove the comment marker to enable the middle rule and apply the configuration again.  I experience it being broken as I described initially.  Sorry my trimmed-down example didn't include everything... it hadn't occurred to me that it would require a progression to happen.

I'll delete that bucket now so you should be able to use the configuration unaltered (unless someone else gets to it first).

**Comment by justinretzolk** [comment]
Hey @akloss-cibo 👋 Thanks for digging into that a bit more! I believe I know what's going on here with that in mind.

The first bit is around the API versioning you'd touched on. There is _something_ going on to that effect, which is touched on in the [first example](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket_lifecycle_configuration#with-neither-a-filter-nor-prefix-specified) of the resource documentation. There's additional notes around this in the [resource schema](https://github.com/hashicorp/terraform-provider-aws/blob/57fa8cd75ba54484ca22992efd02cb3345c9a16b/internal/service/s3/bucket_lifecycle_configuration.go#L95-L98) that I found to be helpful.

In addition to those bits of documentation, I found a [comment by one of our former maintainers](https://github.com/hashicorp/terraform-provider-aws/issues/23299#issuecomment-1048167955) that I think did a pretty good job of explaining what was happening on an older issue. I'd recommend taking a look over that comment as well to try to clear up the configuration.

If you have any questions after reviewing those two items, please do let me know and I'd be happy to continue to discuss it.

**Comment by akloss-cibo** [comment]
I kinda get how/why it happens; my primary concern is that this creates a potentially very destructive situation.  In my sad tale, S3 deleted about 2 billion objects from a bucket because the filter wasn't applied by a lifecycle policy rule... it seems Terraform shouldn't silently allow live configuration to mismatch the plan... well, ever... but especially not in a way as potentially destructive as this.  I'd like to save some other poor soul the same experience of removing 2 billion delete markers.

**Comment by justinretzolk** [comment]
I definitely see where you're coming from @akloss-cibo. Given the scenario you experienced, I wanted try to build a better  understanding why this is the direction that we took, so I dug a bit further through the changes to try to figure it out and found that it links back to #23730. In this issue, we discovered that the way that v3 handled things was not quite right. To quote [one of the comments](https://github.com/hashicorp/terraform-provider-aws/issues/23730#issuecomment-1071188826) from that issue:

> In some sense, v3.x was incorrect in always sending a Prefix within the Filter b/c the AWS S3 API allows users to create rules with `Entire Bucket` scope which would be the empty `Filter` object.

However, given that this was discovered after version 4.5.0 had been released (and during the time period where changes like this were still being backported to the 3.x releases), we needed to consider our [backwards compatibility promise](https://hashicorp.github.io/terraform-provider-aws/faq/#backward-compatibility-promise) of not introducing any major breaking changes in minor versions (something that correcting this behavior would be). To avoid introducing a breaking change, we opted to [document the current behavior](https://github.com/hashicorp/terraform-provider-aws/pull/23746) and note that this would change in the next major version of the provider, when the `prefix` argument is removed (something that will already be a breaking change).

I realize this doesn't change the fact that you experienced behavior that caused you trouble, but I hope that it helps in some way.
